File: backend/app/models/schema.py
"""
Schema definition models for product fields and metrics
"""
from pydantic import BaseModel, Field
from typing import Literal, Optional, List


class FieldDefinition(BaseModel):
    """Definition of a product field"""
    name: str = Field(..., description="Field name (used as key in data JSON)")
    type: Literal["text", "integer", "decimal", "boolean"] = Field(..., description="Field data type")
    unit: Optional[str] = Field(None, description="Unit of measurement (e.g., 'USD', 'W', 'lbs')")
    label: str = Field(..., description="Display label for the field")
    compareDirection: Literal["lower", "higher"] = Field(..., description="Direction for comparison (lower/higher is better)")
    required: bool = Field(True, description="Whether field is required")


class MetricDefinition(BaseModel):
    """Definition of a calculated metric"""
    name: str = Field(..., description="Metric name")
    formula: str = Field(..., description="Formula expression (e.g., 'price / wattage')")
    label: str = Field(..., description="Display label for the metric")
    compareDirection: Literal["lower", "higher"] = Field(..., description="Direction for comparison")
    format: Optional[str] = Field(None, description="Format type (e.g., 'currency', 'percentage')")


class ProductSchema(BaseModel):
    """Product schema definition"""
    fields: List[FieldDefinition] = Field(..., description="List of field definitions")
    metrics: Optional[List[MetricDefinition]] = Field(default_factory=list, description="List of metric definitions")

File: frontend/src/types/schema.ts
export type FieldType = 'text' | 'integer' | 'decimal' | 'boolean'
export type CompareDirection = 'lower' | 'higher'

export interface FieldDefinition {
  name: string
  type: FieldType
  unit?: string
  label: string
  compareDirection: CompareDirection
  required: boolean
}

export interface MetricDefinition {
  name: string
  formula: string
  label: string
  compareDirection: CompareDirection
  format?: string
}

export interface ProductSchema {
  fields: FieldDefinition[]
  metrics?: MetricDefinition[]
}

export interface Product {
  id: string
  sku?: string
  name: string
  product_type: string
  schema: ProductSchema
  data: Record<string, any>
  user_id: string
  created_at: string
  updated_at: string
}

export interface CompetitorListing {
  id: string
  my_product_id: string
  url: string
  retailer_name: string
  product_name: string
  data: Record<string, any>
  last_crawled_at?: string
  created_at: string
}

export interface ProductTemplate {
  id: string
  name: string
  schema: ProductSchema
  is_system: boolean
  user_id?: string
  created_at: string
}

